std/assembly/util/sort.ts

```diff
@@ -43,100 +56,241 @@ export function COMPARATOR<T>(): (a: T, b: T) => i32 {
   }
 }
 
-// @ts-ignore: decorator
-@inline
+// Power Sort implementation (stable) from paper "Nearly-Optimal Mergesorts"
+// https://arxiv.org/pdf/1805.04154.pdf
+// This method usually outperform TimSort.
 export function SORT<T>(
   dataStart: usize,
   length: i32,
   comparator: (a: T, b: T) => i32
 ): void {
-  if (isReference<T>()) {
-    // TODO replace this to faster stable sort (TimSort) when it implemented
-    insertionSort<T>(dataStart, length, comparator);
-  } else {
-    if (length < 256) {
-      insertionSort<T>(dataStart, length, comparator);
-    } else {
-      weakHeapSort<T>(dataStart, length, comparator);
+  var hi = length - 1, n = length;
+  if (n <= INSERTION_SORT_THRESHOLD) {
+    if (ASC_SHRINK_LEVEL < 1) {
+      switch (n) {
+        case 3: {
+          let a = load<T>(dataStart, 1 << alignof<T>());
+          let b = load<T>(dataStart, 2 << alignof<T>());
+          let c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 1 << alignof<T>());
+          b = select<T>(b, a, c >>> 31);
+          a = load<T>(dataStart, 0);
+          c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 0);
+          store<T>(dataStart, select<T>(b, a, c >>> 31), 2 << alignof<T>());
+        }
+        case 2: {
+          let a = load<T>(dataStart, 0);
+          let b = load<T>(dataStart, 1 << alignof<T>());
+          let c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 0);
+          store<T>(dataStart, select<T>(b, a, c >>> 31), 1 << alignof<T>());
+        }
+        case 1:
+        case 0: return;
+      }
     }
+    insertionSort<T>(dataStart, 0, hi, 0, comparator);
+    return;
   }
-}
 
-function insertionSort<T>(
-  dataStart: usize,
-  length: i32,
-  comparator: (a: T, b: T) => i32
-): void {
-  for (let i = 0; i < length; i++) {
-    let a: T = load<T>(dataStart + (<usize>i << alignof<T>())); // a = arr[i]
-    let j = i - 1;
-    while (j >= 0) {
-      let b: T = load<T>(dataStart + (<usize>j << alignof<T>())); // b = arr[j]
-      if (comparator(a, b) < 0) {
-        store<T>(dataStart + (<usize>(j-- + 1) << alignof<T>()), b); // arr[j + 1] = b
-      } else break;
-    }
-    store<T>(dataStart + (<usize>(j + 1) << alignof<T>()), a); // arr[j + 1] = a
+  var lgPlus2         = log2u(n) + 2;
+  var lgPlus2Size     = lgPlus2 << alignof<u32>();
+  var leftRunStartBuf = __alloc(lgPlus2Size);
+  var leftRunEndBuf   = __alloc(lgPlus2Size);
```

Comment:
Can this share a single allocation of double the size? Something like:

```suggestion
  var leftRunStartBuf = __alloc(2 * lgPlus2Size);
  var leftRunEndBuf   = leftRunStartBuf + lgPlus2Size;
```